src/utils/tokens/pickling.ts

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2016 Aviral Dasgupta
+Copyright 2016 OpenMarket Ltd
+Copyright 2018 New Vector Ltd
+Copyright 2020, 2024 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { encodeUnpaddedBase64 } from "matrix-js-sdk/src/matrix";
+import { logger } from "matrix-js-sdk/src/logger";
+
+/**
+ * Retrieves additional data for a pickle key (used to decrypt access tokens and similar
+ * stored in IndexedDB).
+ *
+ * @param {string} userId The user ID who owns the pickle key.
+ * @param {string} deviceId The device ID which owns the pickle key.
+ * @return {Uint8Array} The additional data as a Uint8Array.
+ */
+export function getPickleAdditionalData(userId: string, deviceId: string): Uint8Array {
+    const additionalData = new Uint8Array(userId.length + deviceId.length + 1);
+    for (let i = 0; i < userId.length; i++) {
+        additionalData[i] = userId.charCodeAt(i);
+    }
+    additionalData[userId.length] = 124; // "|"
+    for (let i = 0; i < deviceId.length; i++) {
+        additionalData[userId.length + 1 + i] = deviceId.charCodeAt(i);
+    }
+    return additionalData;
+}
+
+/**
+ * Assuming `data` is fully provided, this converts the data into a usable key then base64 encodes it for use in
+ * other functions. `data` may be partially or completely undefined to deduplicate some validation logic shared
+ * across callers - when undefined in part or in full, this function returns undefined too. If crypto functions
+ * are not available, this function returns undefined regardless of input.
```

Comment:
I think it's important to start documentation with a single-sentence summary of what the method does. Which this kinda does, but if I only read the first sentence, I'm kinda left expecting that the function will do something :sparkles:magical:sparkles: to populate the data. (Also, that first sentence should be a paragraph on its own).

Also also: surprised not to see the word "decrypt" in this description.

```suggestion
 * Decrypts the provided data into a pickle key and base64-encodes it ready for use elsewhere.
 *
 * If `data` is undefined in part or in full, returns undefined.
 *
 *  If crypto functions are not available, returns undefined regardless of input.
```